bib/cmd/bootc-image-builder/image.go

```diff
@@ -154,6 +156,195 @@ func pipelinesForDiskImage(c *ManifestConfig, rng *rand.Rand) (image.ImageKind,
 	return img, nil
 }
 
+func pipelinesForISO(c *ManifestConfig, rng *rand.Rand) (image.ImageKind, error) {
+	if c.Imgref == "" {
+		fail("pipeline: no base image defined")
+	}
+	ref := "ostree/1/1/0"
+	tlsVerify := true
+	containerSource := container.SourceSpec{
+		Source:    c.Imgref,
+		Name:      c.Imgref,
+		TLSVerify: &tlsVerify,
+	}
+
+	img := image.NewAnacondaContainerInstaller(containerSource, ref)
+	img.ExtraBasePackages = rpmmd.PackageSet{
+		Include: []string{
+			"anaconda-dracut",
+			"atheros-firmware",
+			"brcmfmac-firmware",
+			"curl",
+			"dracut-config-generic",
+			"dracut-network",
+			"hostname",
+			"iwlwifi-dvm-firmware",
+			"iwlwifi-mvm-firmware",
+			"kernel",
```

Comment:
I don't understand why we have hardcoded "RPM package lists in Go code" here...what do we have in this that isn't part of https://src.fedoraproject.org/rpms/anaconda/blob/rawhide/f/anaconda.spec#_230 ?

There's another really big topic here which is in the *general* case...I think we need to be using the dependencies from the container.  Especially the kernel package - we want to support people adding e.g. 3rd party hardware drivers into their container build, and the ISO generated from that should use the same kernel, so that they can enable the NIC or storage.

I'm aware there's a lot of history here and we're reusing a existing logic, and that makes sense in the short term.

But the problem I'm talking about here I think is going to drive us towards an architecture where generating the ISO looks more like:

```dockerfile
FROM <provided base image>
RUN dnf -y install anaconda-install-img
```
(Plus then some further logic to convert *that* filesystem tree into an ISO)

Reply:
The kernel and firmware and kernel args being the same for the ISO as for the base image is an understandable requirement (the machine that is going to run the image will have to be able to run the ISO as well), but do we really want to build the ISO from the base image directly, with all the things that the user might have put in their custom images?

It seems like there's so much that can go wrong there.  Packages (a user package that might conflict with an Anaconda dependency, for example), users and groups, services, are all things that can interfere with the installer environment.

Or is this not as likely and big a concern as I'm thinking?

Reply:
Well, I think that the base image doesn't need to be the payload container - but it can be a customer's "golden base image" that e.g. includes drivers for their environment.

I actually think that distributing the installer tree as a container image isn't a bad idea. I'm imagining something like this:

```
bootc-image-builder --type iso --iso-source ghcr.io/example/installer ghcr.io/example/payload
```

`ghcr.io/example/installer` would be defined in a similar way as Colin suggested. Alternatively, a distribution might just provide a base image if building one by users is too hard

Reply:
> But using the user's payload would likely cause more issues than it would solve.

I would definitely say there are not-small ramifications; one obvious one to me is that the user would need to be aware that any custom systems units would run by default in the installer environment too by default, so they'd need to conditionalize...unless we did some work to have the Anaconda environment explicitly boot into a non-default system target.

But what issues specifically were you thinking of?

I know for sure though that the use case of a customized kernel is a real one for some important OCP customers for example, so if we *don't* use the customer base image for the ISO we'd need to in the general case expose some nontrivial subset of the customization functionality for the install environment too.  Maybe just kernel but there may in some cases be userspace tooling and things like "check in after install but before reboot" systemd units.

Reply:
> one obvious one to me is that the user would need to be aware that any custom systems units would run by default in the installer environment too by default, so they'd need to conditionalize...unless we did some work to have the Anaconda environment explicitly boot into a non-default system target.
> 
> But what issues specifically were you thinking of?

That's definitely one of them, potentially the most important one.  There's also packages.  A user package that might conflict with an Anaconda dependency, for example, or .  I just think that if we want our installer to inherit something from the payload container, we can figure out a way to take just the things we need (firmware, kernel modules, kernel args) rather than start from the container and chip away at it.